### ISSUE TITLE
POST on existing resource returning status 400 instead 409

### ISSUE THREAD

**Issue by zopyx** [plone/plone.restapi#740]
Situation: 

Using POST on the root of a Plone site for creating a folder with id `en`. The folder `en` already exists. 

Result:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 142, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 295, in publish_module
  Module ZPublisher.WSGIPublisher, line 229, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 57, in call_object
  Module plone.rest.service, line 23, in __call__
  Module plone.restapi.services, line 19, in render
  Module plone.restapi.services.content.add, line 83, in reply
  Module plone.restapi.services.content.utils, line 94, in add
  Module Products.BTreeFolder2.BTreeFolder2, line 435, in _setObject
  Module plone.folder.ordered, line 259, in _checkId
  Module Products.CMFCore.PortalFolder, line 318, in _checkId
  Module OFS.ObjectManager, line 137, in checkValidId
zExceptions.BadRequest: The id "en" is invalid - it is already in use.

```

which leads to status code 400 on the client side.

400 means "400 => "The request could not be understood by the server due to malformed syntax"

400 is not very much appropriate here.

Perhaps HTTP 409 (Conflict) is more suited.

At least plone.restapi should perform a check by itself in advance before trying to create a new resource if the related `id` already exists and issue a better error.

**Comment by lukasgraf** [comment]
I agree this should be handled better, I might look at this during the sprint.

But I'd focus on catching this condition and returning a proper, JSON formated body. I don't quite agree that 409 would be any better.

> 400 means "400 => "The request could not be understood by the server due to malformed syntax"

No, that phrasing is from an obsolete RFC. The current [RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.5.1) phrasing is much more broad:

> The 400 (Bad Request) status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error

`409` is very specific, and I've only ever seen it with things like WebDAV, where it is used to signal things like merge conflicts during versioning. So I would simply stick with `400`, but with a clear and machine readable message in the body.

**Comment by zopyx** [comment]
> `409` is very specific, and I've only ever seen it with things like WebDAV, where it is used to signal things like merge conflicts during versioning. So I would simply stick with `400`, but with a clear and machine readable message in the body.

👍

**Comment by buchi** [comment]
IMO the error message is already good enough.

With a non-manager user you get:

```json
{
    "message": "The id \"en\" is invalid - it is already in use.",
    "type": "BadRequest"
}
```

When having "Manage portal" permission, the traceback is included additionally.
Maybe we should avoid including a traceback, as it signals an unhandled exception.
Currently there's now way to raise an exception without including the traceback for users with "Manage portal" permission.
We need to raise an exception to ensure the transaction doesn't get committed and to break out easily of the control flow.

Solving #165 should provide a way to raise exceptions for handled errors which do not include a traceback.

**Comment by jensens** [comment]
Closing as outdated (6 years old). Discussion concluded: HTTP 400 is correct per RFC 7231.
If you think this issue is still relevant please reopen it. Thank you for the original report!
